File: backend.scala/app/com/skewwhiffy/auraltester/internalnotation/InternalNotationFactory.scala
package com.skewwhiffy.auraltester.internalnotation

import com.skewwhiffy.auraltester.clefs.{Clef, ClefFactory}
import com.skewwhiffy.auraltester.notes.{AbsoluteNote, DirectedInterval}
import com.skewwhiffy.auraltester.scales.Key

import javax.inject.Inject

class InternalNotationFactory @Inject()(
  clefFactory: ClefFactory,
  intervalFactory: IntervalFactory,
  keyFactory: KeyFactory,
  noteFactory: NoteFactory
) {
  def clef(clefRaw: String): Clef = clefRaw.toLowerCase match {
    case "treble" => clefFactory.treble
    case "alto" => clefFactory.alto
    case "tenor" => clefFactory.tenor
    case "bass" => clefFactory.bass
    case _ => throw new IllegalArgumentException(s"Unrecognized clef type: '$clefRaw'")
  }

  def getNote(noteRaw: String): AbsoluteNote = noteFactory.getAbsoluteNote(noteRaw)

  def getNotes(notesRaw: String): List[AbsoluteNote] = notesRaw
    .split(' ')
    .map(getNote)
    .toList

  def getDirectedInterval(rawInterval: String): DirectedInterval = intervalFactory
    .getDirectedInterval(rawInterval)

  def getDirectedIntervals(rawIntervals: String): List[DirectedInterval] = intervalFactory
    .getDirectedIntervals(rawIntervals)

  def getKey(rawKey: String): Key = keyFactory.getKey(rawKey)
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/factory/InternalNotationFactory.kt
package com.skewwhiffy.notation.factory

import com.skewwhiffy.notation.model.note.AbsoluteNote
import com.skewwhiffy.notation.model.clef.Clef
import com.skewwhiffy.notation.model.interval.DirectedInterval
import org.springframework.stereotype.Service
import java.util.*

@Service
class InternalNotationFactory(
  private val clefFactory: ClefFactory,
  val intervalFactory: IntervalFactory,
  val keyFactory: KeyFactory,
  val noteFactory: NoteFactory
) {
  fun clef(clefRaw: String): Clef {
    return when (clefRaw.lowercase(Locale.UK)) {
      "treble" -> clefFactory.treble
      "alto" -> clefFactory.alto
      "tenor" -> clefFactory.tenor
      "bass" -> clefFactory.bass
      else -> throw IllegalArgumentException("Unrecognized clef type: '$clefRaw'")
    }
  }

  fun getNote(noteRaw: String): AbsoluteNote {
    return noteFactory.getAbsoluteNote(noteRaw)
  }

  fun getNotes(notesRaw: String): List<AbsoluteNote> = notesRaw
    .split(' ')
    .map(::getNote)
    .toList()

  fun getDirectedInterval(rawInterval: String): DirectedInterval = intervalFactory
    .getDirectedInterval(rawInterval)

  /*
  def getDirectedIntervals(rawIntervals: String): List[DirectedInterval] = intervalFactory
    .getDirectedIntervals(rawIntervals)

  def getKey(rawKey: String): Key = keyFactory.getKey(rawKey)
}

   */
}